File: samples/monovo/cpp/src/monovo.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <thread>

#include <cassert>
#include <filesystem>

#include <opencv2/opencv.hpp>

bool checkFileExistence(const std::string &str)
{
    std::ifstream ifs(str);
    return ifs.is_open();
}

std::vector<std::vector<cv::DMatch>> refineMatches(std::vector<std::vector<cv::DMatch>> matches)
{
    std::vector<std::vector<cv::DMatch>> refMatches;
    for (auto match : matches)
    {
        if (match[0].distance < 0.75 * match[1].distance)
        {
            refMatches.push_back(match);
        }
    }
    return refMatches;
}

std::vector<std::string> getFileList(std::string path)
{
    std::vector<std::string> list;
    for (auto &entry : std::filesystem::directory_iterator(path))
    {
        list.push_back(entry.path().filename().string());
    }
    return list;
}

int main(int argc, char *argv[])
{
    // initailize orb and bfmatcher
    cv::Ptr<cv::FeatureDetector> orb = cv::ORB::create(500, 1.2f, 8, 31, 0, 2, cv::ORB::HARRIS_SCORE, 31, 20);
    cv::Ptr<cv::BFMatcher> bfMatcher = cv::BFMatcher::create(cv::NORM_HAMMING, false);

    // Camera parameter
    double fx = 540.627;
    double fy = 550.577;
    double cx = 320.833;
    double cy = 240.796;
    double k1 = 0.1546;
    double k2 = -0.331;
    double p1 = -0.00123;
    double p2 = -0.0001;
    cv::Mat K = (cv::Mat1d(3, 3) << fx, 0, cx, 0, fy, cy, 0, 0, 1);
    cv::Mat distCoeffs = (cv::Mat1d(1, 4) << k1, k2, p1, p2);

    std::string image_dir = "images/";
    std::vector<std::string> fsList = getFileList(image_dir);

    cv::Mat last_t, last_R;
    bool first = true;

    for (int i = 0; i < fsList.size() - 1; i++)
    {
        cv::Mat image1 = cv::imread(image_dir + fsList[i]);
        cv::Mat image2 = cv::imread(image_dir + fsList[i + 1]);

        // detect ORB keypoint and compute description
        std::vector<cv::KeyPoint> kp1, kp2;
        cv::Mat desc1, desc2;
        std::vector<std::vector<cv::DMatch>> matches;
        orb->detectAndCompute(image1, cv::Mat(), kp1, desc1);
        orb->detectAndCompute(image2, cv::Mat(), kp2, desc2);

        // brute force knn match
        bfMatcher->knnMatch(desc1, desc2, matches, 2);

        // apply Lowe's test
        matches = refineMatches(matches);

        // draw match
        cv::Mat drawImg;
        cv::drawMatches(image1, kp1, image2, kp2, matches, drawImg);

        std::vector<cv::Point2f> mp1, mp2;
        for (auto m : matches)
        {
            mp1.push_back(kp1[m[0].queryIdx].pt);
            mp2.push_back(kp2[m[0].trainIdx].pt);
        }

        // calculate essential matrix and decompose to R and t
        cv::Mat E, R, t, mask;
        E = cv::findEssentialMat(mp2, mp1, K, cv::RANSAC, 0.999, 1.0, mask);
        cv::recoverPose(E, mp2, mp1, K, R, t, mask);
        if (first)
        {
            first = false;
            last_t = t.clone();
            last_R = R.clone();
        }
        cv::Mat Tw = last_t + last_R * t;
        cv::Mat Rw = R * last_R;
        std::cout << Rw << std::endl << Tw << std::endl;

        // update last trajectories
        last_t = Tw.clone();
        last_R = Rw.clone();

        cv::imshow("image", drawImg);
        if (cv::waitKey(1) == 27)
        {
            break;
        }
    }
    cv::destroyAllWindows();
    return 0;
}

File: samples/monovo/python/monovo.py
import os
import time
import glob
import collections

import cv2
import numpy as np

def refineMatches(matches):
    refMatches = []
    for m, n in matches:
        if m.distance < 0.75 * n.distance:
            refMatches.append([m])
    return refMatches


if __name__ == '__main__':
    # Camera parameter
    fx = 540.627
    fy = 550.577
    cx = 320.833
    cy = 240.796
    k1 = 0.1546
    k2 = -0.331
    p1 = -0.00123
    p2 = -0.0001

    K = np.identity(3)
    K[0, 0] = fx
    K[1, 1] = fy
    K[0, 2] = cx
    K[1, 2] = cy
    distCoeffs = (k1, k2, p1, p2)

    # initialize orb and bfmatcher
    orb = cv2.ORB_create(nfeatures=500, scaleFactor=1.2, nlevels=8, edgeThreshold=31,
                         firstLevel=0, WTA_K=2, scoreType=cv2.ORB_HARRIS_SCORE, patchSize=31, fastThreshold=20)
    bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=False)

    isFirst = True
    last_t = [0, 0, 0]
    last_R = np.identity(3)
    last_R[0,0] = 1
    last_R[1,1] = 1
    last_R[2,2] = 1
    # input file list
    image_list = 'images/*.jpg'
    flist = [name for name in sorted(glob.glob(image_list))]
    
    for i in range(0, len(flist)-1):
        image1 = cv2.imread(flist[i])
        image2 = cv2.imread(flist[i+1])
        # detect ORB keypoint and compute description
        kp1, desc1 = orb.detectAndCompute(image1, None)
        kp2, desc2 = orb.detectAndCompute(image2, None)
        # brute force knn match
        matches = bf.knnMatch(desc1, desc2, k=2)
        matches = refineMatches(matches)
        # draw match
        drawImg = cv2.drawMatchesKnn(image1, kp1, image2, kp2, matches, None)
        mp1 = []
        mp2 = []
        for m in matches:
            mp1.append(kp1[m[0].queryIdx].pt)
            mp2.append(kp2[m[0].trainIdx].pt)
        mp1 = np.array([mp1]).reshape(len(mp1), 2)
        mp2 = np.array([mp2]).reshape(len(mp2), 2)
        # Calculate Essential Matrix and Decompose to R,t
        E, e_inliers = cv2.findEssentialMat(points1=mp2, points2=mp1, cameraMatrix=K, method=cv2.RANSAC, prob=0.999, threshold=1.0)
        inliercnt, R, t, inliers = cv2.recoverPose(E, mp2, mp1, K)
        t = t.flatten()
        if isFirst:
            isFirst = False
            last_R = R
            last_t = t

        mulMat = last_R @ t
        Tw  = last_t + (last_R @ t)
        Rw = R @ last_R
        print(Rw)
        print(Tw)

        # update last trajectories
        last_t = Tw
        last_R = Rw

        cv2.imshow("image", drawImg)
        if cv2.waitKey(1) == 27:
            break

    cv2.destroyAllWindows()
